File: solved/135.Candy.py
class Solution(object):
  def candy(self, ratings):
    """
    :type ratings: List[int]
    :rtype: int
    """
    N = len(ratings)
    print(ratings)
    candies = [1] * N

    for i in range(1, N):
      if ratings[i] > ratings[i-1]:
        candies[i] = candies[i-1]+1

    for i in range(N-2, -1, -1):
      if ratings[i] > ratings[i+1]:
        if not candies[i] > candies[i+1]:
          candies[i] = candies[i+1]+1
    
    print(candies)
    res = sum(candies)
    print(res)
    return res

Solution.candy(0, [1,0,2]) # 5
Solution.candy(0, [1,2,2]) # 4
Solution.candy(0, [1,3,2,2,1]) # 7
Solution.candy(0, [1,2,87,87,87,2,1]) # 13
Solution.candy(0, [1,3,4,5,2]) # 11

File: solved/135.Candy.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        int N = ratings.size();
        vector<int> res(N, 1);

        for (int i = 1; i < N; i++) {
            if (ratings[i] > ratings[i-1]) {
                res[i] = res[i-1]+1;
            }
        }

        int ress = 0;
        for (int i = N-1; i > 0; i--) {
            if (ratings[i] < ratings[i-1] && !(res[i] < res[i-1])) {
                res[i-1] = res[i]+1;
            }
        }

        return accumulate(res.begin(), res.end(), 0);
    }
};
